### PR TITLE
Add Individual Files per Entries Option

### LINKED ISSUE #10: Add keys to collection of each content_type
Currently, each content_type is placed in an collection, but without any keys. Would it be possible to get the collection indexed by entry title, so that we can cherry pick a particular entry?

### LINKED ISSUE #25: Feature Request: Generate separate markdown files per entry
The existing importer is great and works fine as it is. But for creating a more flexible website based on Contentful data, some features are not possible by a single large data yaml file. Specially it is a URL generation by Jekyll (without special plugins).

Therefor I have something like that in mind:

* configure per content type if it should be rendered in a large yaml data file or in separate markdown files
* for markdownfiles define
** location to same the files, e.g. /teammembers/_posts/
** Contentful field for usage as markdown content
** Contentful field for the date within the markdown filename
** template key/value pairs for yaml (e.g. Jekyll typical layout property)
** simplification for references used as Jekyll category and tag (e.g. by a configuration of Contenful fields used for tags and categories and than reduce the reference entries to eliminate ids)

### PR DESCRIPTION
**Pull request by dlitvakb**
Fixes #23 
Fixes #10 
Fixes #25

### CODE REVIEW

File: lib/jekyll-contentful-data-import/single_file_data_exporter.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+require 'jekyll-contentful-data-import/base_data_exporter'
+
+module Jekyll
+  module Contentful
+    # Single File Data Exporter Class
+    #
+    # Serializes Contentful data into a Single YAML File
```
Comment (mariobodemann) [inline lib/jekyll-contentful-data-import/single_file_data_exporter.rb]: typo: `S`ingle vs `s`ingle and `F`ile vs `f`ile.

File: spec/jekyll-contentful/multi_file_data_exporter_spec.rb [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+describe Jekyll::Contentful::MultiFileDataExporter do
+  subject { described_class.new('foo', []) }
+
+  describe 'instance methods' do
+    describe '#base_directory' do
+      it 'default directory' do
+        expect(subject.base_directory).to eq(Dir.pwd)
+      end
+
+      it 'overriden directory' do
```
Comment (mariobodemann) [inline spec/jekyll-contentful/multi_file_data_exporter_spec.rb]: overri`d`en vs overri`dd`en.

File: spec/jekyll-contentful/multi_file_data_exporter_spec.rb [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    describe '#setup_directory' do
+      it 'default directory' do
+        expected = File.join(Dir.pwd, '_data', 'contentful', 'spaces')
+        expect(FileUtils).to receive(:mkdir_p).with(expected)
+
+        subject.setup_directory(expected)
+      end
+
+      it 'overridden directory' do
```
Comment (mariobodemann) [inline spec/jekyll-contentful/multi_file_data_exporter_spec.rb]: :+1:

File: spec/jekyll-contentful/multi_file_data_exporter_spec.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      end
+
+      it 'does nothing with no entries' do
+        subject.run
+      end
+
+      it 'serializes entries' do
+        expect_any_instance_of(::Jekyll::Contentful::Serializer).to receive(:serialize) { {} }
+
+        subject.run
```
Comment (mariobodemann) [inline spec/jekyll-contentful/multi_file_data_exporter_spec.rb]: Hmm, that style of testing only validates, that no error/exception is thrown. Do you mind adding some more precise validations? Aka that the right entry got serialized to the right result?

File: spec/jekyll-contentful/multi_file_data_exporter_spec.rb [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+      end
+
+      it 'creates a file per entry' do
+        subject = described_class.new('foo', [EntryDouble.new('bar', ContentTypeDouble.new('bar_ct'))])
+
+        expected_directory_path = File.join(Dir.pwd, '_data', 'contentful', 'spaces', 'foo', 'bar_ct')
+        expect(FileUtils).to receive(:mkdir_p).with(expected_directory_path)
+        expect(File).to receive(:open).with(File.join(expected_directory_path, 'bar.yaml'), 'w')
```
Comment (mariobodemann) [inline spec/jekyll-contentful/multi_file_data_exporter_spec.rb]: That does not test, what the name suggest, right? I was expecting it to check the existence of all the files created, and not just one. Or I'm miss reading the test name ... :)

Comment (dlitvakb) [inline spec/jekyll-contentful/multi_file_data_exporter_spec.rb]: There is a single entry, so it just checks for a single file.

File: spec/jekyll-contentful/multi_file_data_exporter_spec.rb (hunk 4 above)
Comment (dlitvakb) [inline spec/jekyll-contentful/multi_file_data_exporter_spec.rb]: There are tests for the serializers just doing that, this here tests that the logic is followed

Comment (mariobodemann) [inline spec/jekyll-contentful/multi_file_data_exporter_spec.rb]: Sorry, but I do not see how you test that the logic is followed. The only thing I can see is validating that nothing throws, unless I'm overlooking something.

File: spec/jekyll-contentful/multi_file_data_exporter_spec.rb (hunk 5 above)
Comment (mariobodemann) [inline spec/jekyll-contentful/multi_file_data_exporter_spec.rb]: Then, I'm confused by the naming ... ;) Should we call it creates a file or rather have the test create multiple files and validate that?

Comment (dlitvakb) [inline spec/jekyll-contentful/multi_file_data_exporter_spec.rb]: I'm testing that when running the exporter it's calling the serializer.

File: spec/jekyll-contentful/multi_file_data_exporter_spec.rb (hunk 4 above)
Comment (dlitvakb) [inline spec/jekyll-contentful/multi_file_data_exporter_spec.rb]: I'm testing that when the exporter is running it's calling the serializer.

File: spec/jekyll-contentful/multi_file_data_exporter_spec.rb (hunk 5 above)
Comment (dlitvakb) [inline spec/jekyll-contentful/multi_file_data_exporter_spec.rb]: If you look at the setup for the `subject` I'm sending it a list with a single entry, therefore creating a single file for that entry is the expected behaviour. As it is creating a file for a specific entry with a path that looks like `{base_dir}/_data/contentful/spaces/{space_alias}/{content_type_id}/{entry_id}.yaml`. While the `single_file_data_exporter` creates a file for the space instead (and all content types and entries are inside that file).

**Pull request merged**